src/nl/hannahsten/texifyidea/refactoring/inlinecommand/LatexInlineCommandProcessor.kt

```diff
@@ -0,0 +1,124 @@
+import com.intellij.usageView.UsageViewDescriptor
+import nl.hannahsten.texifyidea.psi.LatexCommands
+import nl.hannahsten.texifyidea.psi.LatexKeyvalPair
+import nl.hannahsten.texifyidea.psi.LatexPsiHelper
+import nl.hannahsten.texifyidea.psi.LatexRequiredParamContent
+import nl.hannahsten.texifyidea.util.childrenOfType
+import nl.hannahsten.texifyidea.util.definitionCommand
+
+/**
+ * Recieves settings and a target and performs the requested refactoring
+ *
+ * @see com.intellij.refactoring.inline.InlineMethodProcessor
+ *
+ * @author jojo2357
+ */
+class LatexInlineCommandProcessor(
+    myProject: Project,
+    private val inlineCommand: LatexCommands,
+    private val originalReference: PsiElement?,
+    private val isInlineThisOnly: Boolean,
+    private val isKeepTheDeclaration: Boolean,
+    private val myScope: SearchScope = GlobalSearchScope.projectScope(myProject)
+) : BaseRefactoringProcessor(myProject) {
+
+    override fun createUsageViewDescriptor(usages: Array<out UsageInfo>): UsageViewDescriptor {
+        return LatexInlineCommandDescriptor(inlineCommand)
+    }
+
+    override fun findUsages(): Array<UsageInfo> {
+        val tempreferences =
+            ReferencesSearch.search(inlineCommand.definitionCommand()!!, myScope).findAll().asSequence()
+
+        return tempreferences
+            .distinct()
+            .map(PsiReference::getElement)
+            .map(::UsageInfo)
+            .toList()
+            .toTypedArray()
+    }
+
+    override fun performRefactoring(usages: Array<out UsageInfo>) {
+        if (isInlineThisOnly) {
+            if (originalReference != null)
+                replaceUsage(originalReference)
+            else if (!ApplicationManager.getApplication().isUnitTestMode)
+                throw IllegalStateException("Inline this requested with no original reference (" + usages.size + ")")
+            // else we are probably trying to elicit no response because this is illegal
+        }
+        else if (usages.isNotEmpty()) {
+            ApplicationManager.getApplication().runWriteAction {
+                CommandProcessor.getInstance().executeCommand(myProject, {
+                    for (replaceUsage in usages) {
+                        val replacereference = replaceUsage.element ?: continue
+                        replaceUsage(replacereference)
+                    }
+                    if (!isKeepTheDeclaration)
+                        inlineCommand.delete()
+                }, commandName, "Texify")
+            }
+        }
+    }
+
+    /**
+     * Does the lifting by...replacing usages with the entirety of the file to be inlined
+     *
+     * @param psiElement The element to remove and replace with the contents of the file
+     */
+    private fun replaceUsage(psiElement: PsiElement) {
+        val calledRequiredArgs = psiElement.childrenOfType(LatexRequiredParamContent::class).toList()
+        val calledOptionalArgs = psiElement.childrenOfType(LatexKeyvalPair::class).toList()
```

Comment:
To rely less on the exact psi structure, it is better to use LatexCommands.optionalParameterMap

Reply:
why is it a map? I didn't see any non-null values, only keys

Reply:
If you have a command `\mycmd[a=3,b=4]{text}` it's supposed to be a map `{ "a" to 3, "b" to 4}`

Reply:
How would you define that? Never seen a custom command do that

Reply:
Some packages have it like that, don't recall right now which ones but you can check the usages of that method to find examples probably

Reply:
LaTeX is strange. Some packages even thought it was a good idea to use `<...>` for parameters instead of `{...}`! Parser doesn't like that.

Reply:
Well i made it so maps are not supported at this time, and I use the parammap as requested